### PR TITLE
Add diffblue tests [Internal review] 

### PR DESCRIPTION
**Pull request by paulkennethkent**
The test cases contained in this PR were generated by [Diffblue's](http://www.diffblue.com) [Deeptest](http://www.diffblue.com/papers/overview.pdf) software. 

One of our missions is to provide useful tests to open source projects. In return, we would appreciate your feedback about our tests to help us further improve our tool. 
	
Results for various open source projects can be viewed at:

   [http://diff.blue/demo](http://diff.blue/demo)

We welcome any feedback you have on these tests.

### CODE REVIEW

File: librec/src/test/java/librec/data/DenseMatrixTest.java [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+import java.lang.reflect.Array;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+public class DenseMatrixTest {
+
+  @Rule public ExpectedException thrown = ExpectedException.none();
+
+  /* testedClasses: DenseMatrix */
```
Comment (JohnLBergqvist) [inline librec/src/test/java/librec/data/DenseMatrixTest.java]: Remove the `testedClasses` & `Test generated...` comments.

File: librec/src/test/java/librec/data/DenseMatrixTest.java [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+  /*
+   * Test generated by Diffblue Deeptest.
+   * This test case covers the entire method.
+   */
+
+  @Test
+  public void numColumnsOutputZero() {
+
+    // Arrange
+    final DenseMatrix objectUnderTest = new DenseMatrix(null, 0, 0);
```
Comment (JohnLBergqvist) [inline librec/src/test/java/librec/data/DenseMatrixTest.java]: I don't think we'll get edge cases like these accepted.

File: librec/src/test/java/librec/data/DenseMatrixTest.java [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+import java.lang.reflect.Method;
+
+public class DenseMatrixTest {
+
+  @Rule public ExpectedException thrown = ExpectedException.none();
+
+  /* testedClasses: DenseMatrix */
+
+  /*
+   * Test generated by Diffblue Deeptest.
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/data/DenseMatrixTest.java]: Did we decide to eliminate the line coverage comments from tests we plan to submit?

**Review by JohnLBergqvist — CHANGES_REQUESTED** [review]
I think there's far too many tests here. Also, I don't think its worth submitting tests that take in mostly 0 and return 0.

File: librec/src/test/java/librec/data/DenseMatrixTest.java (hunk 3 above)
Comment (paulkennethkent) [reply librec/src/test/java/librec/data/DenseMatrixTest.java]: If so, we should update the diffblue.yml in the cx repo.

File: librec/src/test/java/librec/data/DenseMatrixTest.java [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    Assert.assertEquals(Double.NaN, retval, 0.0);
+  }
+
+  /*
+   * Test generated by Diffblue Deeptest.
+   * This test case covers:
+   *  - conditional line 92 branch to line 92
+   */
+
+  @Test
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/data/DenseMatrixTest.java]: This looks like a useful test.

File: librec/src/test/java/librec/data/DenseMatrixTest.java [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+
+  /*
+   * Test generated by Diffblue Deeptest.
+   * This test case covers:
+   *  - conditional line 343 branch to line 343
+   *  - conditional line 343 branch to line 344
+   */
+
+  @Test
+  public void hadamardProductInputNotNullNotNullOutputException() throws Exception {
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/data/DenseMatrixTest.java]: This looks like a useful test.

File: librec/src/test/java/librec/data/DenseMatrixTest.java [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+  }
+
+  /*
+   * Test generated by Diffblue Deeptest.
+   * This test case covers:
+   *  - conditional line 743 branch to line 744
+   */
+
+  @Test
+  public void inverseOutputRuntimeException() {
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/data/DenseMatrixTest.java]: This looks like a useful test.

File: librec/src/test/java/librec/data/DenseVectorTest.java [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+   * Test generated by Diffblue Deeptest.
+   * This test case covers:
+   *  - conditional line 230 branch to line 230
+   *  - conditional line 230 branch to line 232
+   *  - conditional line 232 branch to line 232
+   *  - librec/data/DenseVector.java:233: loop: 1 iterations
+   */
+
+  @Test
+  public void minusInputNotNullOutputNotNull2() {
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/data/DenseVectorTest.java]: This looks like a useful test.

File: librec/src/test/java/librec/data/DenseVectorTest.java [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+  /*
+   * Test generated by Diffblue Deeptest.
+   * This test case covers:
+   *  - conditional line 189 branch to line 189
+   *  - librec/data/DenseVector.java:191: loop: 1 iterations
+   */
+
+  @Test
+  public void minusInputZeroOutputNotNull() {
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/data/DenseVectorTest.java]: This looks like a useful test in combination with above test.

File: librec/src/test/java/librec/data/RatingContextTest.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
+package librec.data;
+
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/data/RatingContextTest.java]: Possibly consider scrapping this Test.class - the comparison tests are essentially doing x = (a - b), if x > 0, return 1, if x < 0, return -1, if equal, return 0. I don't think there are any edge cases worth highlighting for this functionality. 

The .getMood / .getLocation() / .getTimestamp() / .getAccompany() tests are all trying to return a attribute that hasn't been set in these tests. Unless there are any accmpanying tests where .setMood() ect is called, I think they may be worth dropping.

**Review by petersear — COMMENTED** [review]
Quite a lot of tests!
Can all refs to the following be removed as neither appear to be used in any of the tests?
java.lang.reflect
mockito
If so, does that mean we can removed the pom dependency on deeptestutils?

File: librec/src/test/java/librec/util/GammaTest.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  /* testedClasses: Gamma */
+
+  /*
+   * Test generated by Diffblue Deeptest.
+   * This test case covers:
+   *  - conditional line 68 branch to line 69
+   */
+
+  @Test
+  public void digammaInputNegativeInfinityOutputNaN() {
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/util/GammaTest.java]: Java Double.NEGATIVE_INFINITY != Double.NaN by default. That makes this an interesting and quite possibly useful test case!

File: librec/src/test/java/librec/util/GammaTest.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  /*
+   * Test generated by Diffblue Deeptest.
+   * This test case covers:
+   *  - conditional line 68 branch to line 68
+   *  - conditional line 68 branch to line 72
+   *  - conditional line 72 branch to line 73
+   */
+
+  @Test
+  public void digammaInputZeroOutputNegativeInfinity() {
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/util/GammaTest.java]: I also like this! Are there any digamma tests with inputs that are natural numbers?

File: librec/src/test/java/librec/util/GaussianTest.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+  /* testedClasses: Gaussian */
+
+  /*
+   * Test generated by Diffblue Deeptest.
+   * This test case covers:
+   *  - conditional line 55 branch to line 56
+   */
+
+  @Test
+  public void cdfInputNegativeOutputZero() {
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/util/GaussianTest.java]: In order to get a non-0 or non-1 return value from the Gaussian.cdf() tests we'll need to find an input between -8.0 and 8. Is there anything from Deeptest that demonstrates this?

File: librec/src/test/java/librec/util/KernelSmoothingTest.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
+package librec.util;
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/util/KernelSmoothingTest.java]: KernelSmoothing tests here are actually fantastic. A PR with these alone would be decently impressive.

File: librec/src/test/java/librec/util/ListsTest.java [hunk 14]
```diff
@@ -0,0 +1,661 @@
+package librec.util;
+
+import static org.mockito.AdditionalMatchers.or;
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/util/ListsTest.java]: Mockito imports here don't appear to be necessary

File: librec/src/test/java/librec/util/ListsTest.java [hunk 15]
```diff
@@ -0,0 +1,661 @@
...
+import static org.mockito.Matchers.isA;
+import static org.mockito.Matchers.isNull;
+
+import librec.util.FileIO.Converter;
+import librec.util.Lists;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/util/ListsTest.java]: org.junit.runner.RunWith also seems to be unecessary. (Correct me if I am wrong!)

File: librec/src/test/java/librec/util/MathsTest.java [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+  /* testedClasses: Maths */
+
+  /*
+   * Test generated by Diffblue Deeptest.
+   * This test case covers:
+   *  - conditional line 78 branch to line 79
+   */
+
+  @Test
+  public void factorialInputNegativeOutputZero() {
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/util/MathsTest.java]: This is a cool test. Highlights a possible bug?

File: librec/src/test/java/librec/util/MathsTest.java [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+    Assert.assertEquals(0, retval);
+  }
+
+  /*
+   * Test generated by Diffblue Deeptest.
+   * This test case covers:
+   *  - conditional line 144 branch to line 145
+   */
+
+  @Test
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/util/MathsTest.java]: This is a nice test too

**Review by JohnLBergqvist — APPROVED** [review]

File: librec/src/test/java/librec/util/StringsTest.java [hunk 18]
```diff
@@ -0,0 +1,341 @@
...
+    final int maxLength = 4;
+
+    // Act
+    final String retval = Strings.last(str, maxLength);
+
+    // Assert result
+    Assert.assertEquals("...!", retval);
+  }
+
+  @Test
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/util/StringsTest.java]: :thinking:

File: librec/src/test/java/librec/util/StringsTest.java [hunk 19]
```diff
@@ -0,0 +1,341 @@
...
+    final String str = "5";
+
+    // Act
+    final int retval = Strings.toInt(str);
+
+    // Assert result
+    Assert.assertEquals(5, retval);
+  }
+
+  @Test
```
Comment (sudo-jaa) [inline librec/src/test/java/librec/util/StringsTest.java]: ?!

File: librec/pom.xml [hunk 20]
```diff
@@ -21,6 +21,11 @@
             <artifactId>guava</artifactId>
             <version>16.0</version>
         </dependency>
+        <dependency>
```
Comment (sudo-jaa) [inline librec/pom.xml]: I believe this dependency can be removed.

Comment (petersear) [reply librec/pom.xml]: I think so too. Otherwise, all looks good.

**Review by petersear — APPROVED** [review]
